src/org/thoughtcrime/securesms/loki/views/MessageAudioView.kt

```diff
@@ -0,0 +1,336 @@
+package org.thoughtcrime.securesms.loki.views
+
+import android.content.Context
+import android.content.res.ColorStateList
+import android.graphics.Color
+import android.graphics.PorterDuff
+import android.graphics.drawable.AnimatedVectorDrawable
+import android.util.AttributeSet
+import android.view.View
+import android.view.View.OnTouchListener
+import android.view.ViewGroup
+import android.widget.FrameLayout
+import android.widget.ImageView
+import android.widget.ProgressBar
+import android.widget.TextView
+import androidx.annotation.ColorInt
+import androidx.core.content.ContextCompat
+import androidx.core.graphics.ColorUtils
+import kotlinx.coroutines.*
+import network.loki.messenger.R
+import org.greenrobot.eventbus.EventBus
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode
+import org.thoughtcrime.securesms.ApplicationContext
+import org.thoughtcrime.securesms.attachments.DatabaseAttachment
+import org.thoughtcrime.securesms.audio.AudioSlidePlayer
+import org.thoughtcrime.securesms.components.AnimatingToggle
+import org.thoughtcrime.securesms.database.AttachmentDatabase
+import org.thoughtcrime.securesms.database.DatabaseFactory
+import org.thoughtcrime.securesms.events.PartProgressEvent
+import org.thoughtcrime.securesms.logging.Log
+import org.thoughtcrime.securesms.loki.api.PrepareAttachmentAudioExtrasJob
+import org.thoughtcrime.securesms.mms.AudioSlide
+import org.thoughtcrime.securesms.mms.SlideClickListener
+import java.io.IOException
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class MessageAudioView: FrameLayout, AudioSlidePlayer.Listener {
+
+    companion object {
+        private const val TAG = "AudioViewKt"
+    }
+
+    private val controlToggle: AnimatingToggle
+    private val container: ViewGroup
+    private val playButton: ImageView
+    private val pauseButton: ImageView
+    private val downloadButton: ImageView
+    private val downloadProgress: ProgressBar
+    private val seekBar: WaveformSeekBar
+    private val totalDuration: TextView
+
+    private var downloadListener: SlideClickListener? = null
+    private var audioSlidePlayer: AudioSlidePlayer? = null
+
+    /** Background coroutine scope that is available when the view is attached to a window. */
+    private var asyncCoroutineScope: CoroutineScope? = null
+
+    private val loadingAnimation: SeekBarLoadingAnimation
+
+    constructor(context: Context): this(context, null)
+
+    constructor(context: Context, attrs: AttributeSet?): this(context, attrs, 0)
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
+        View.inflate(context, R.layout.message_audio_view, this)
+        container = findViewById(R.id.audio_widget_container)
+        controlToggle = findViewById(R.id.control_toggle)
+        playButton = findViewById(R.id.play)
```

Comment:
Just for future reference, you don't need to do this in Kotlin :) . You can just reference the view directly without first doing `findViewById(...)`, assigning it, etc.

Reply:
This is also why we have the convention of using lower camel case for view IDs

Reply:
That was done intentionally. Kotlin synthetics are no longer recommended practice by Google https://www.reddit.com/r/androiddev/comments/ala9p2/why_kotlinx_synthetic_is_no_longer_a_recommended/?utm_source=amp&utm_medium=&utm_content=post_num_comments

They show down things a bit and View Binding seems like a good alternative https://developer.android.com/topic/libraries/view-binding

Reply:
Oh okay then, what a Google move haha

Reply:
New day, new library at Google :) once this thing is out of alpha it might become the default xml view markup replacement https://developer.android.com/jetpack/compose